### ISSUE TITLE
Questions & Suggestions For Cloud Stores 

### ISSUE THREAD

**Issue by mayoatte** [tusdotnet/tusdotnet#154]
Hi,

Thanks a ton for all your hard work on this library, I was very glad to find it.

I am integrating the library into our app now and that requires writing several custom stores (TusStore, TusCreationStore, TusExpirationStore). All these stores live in the cloud (blob storage or databases) and if i understand things right, there will be a lot of network calls . I think there might be an opportunity to optimize some of these calls. Let me try to explain:

** Create File Handler **
While handling the Create Intent, we will use the creation store to save the upload length / metadata and the expiration store to save the expiration.  It seems to me that in many cases, these two stores will be the same underlying data store e.g. this library uses the file system for both. In my case, i need to use the same cloud database for both. 

Now, the two stores are called sequentially, one after the other in the create handler. I would love to be able to create the single in memory object that holds the creation and expiration data once  and then save it to the database after the tus handler is done. Right now, each call to either store requires a network call to the database. This gets worse because our system is multitenant and we need some additional info on each call.  Some questions here: 

- What is the DI lifetime used for a store? Is it a singleton? Scoped to the request? Transient? Trying to figure out if we can store some state in the instance of the store.
- What do you think of adding one more callback event after any of the intent handlers is done with processing. This would be safe time to save any data we have accumulated during the request. Something like OnTusHandlerCompleted

** Write File Handler

During this one, we need the normal tus store (for the stream append), the creation store (to get the upload length so we can check if the upload is done or has been exceeded), and we also need the expiration store (to get / set the expiration depending on if it is sliding). Again, we need a variety of information from multiple stores that might be backed by one or more datastores,  all in sequence.  A naïve implementation could result in a lot of network calls for what might be a single db record.

- Being able to cache state in instances might help here (same question as above).
- An event callback like OnBeforeTusHandling could allows us to make a single call to the datastore to fetch the various bits of data we need.

--------
What do you think of all this? IMO, it boils down to being able to treat the handling of a single request in a cohesive way (get data --> process --> save data) despite the (appropriate) separation of the various kinds of stores. Integration with normal DI and additional events may help.

Thanks for listening.

**Comment by mayoatte** [comment]
Another way in which the above could help:

In cases where the client uploads data with the Creation call (Create With Upload). We could conceivably, end up transferring the bytes to a remote store and not have to save or deal with anything else (no metadata, no upload length, no expiration). This savings would be especially high for small files (saves on subsequent processing etc).

**Comment by mayoatte** [comment]
I found the lifetime in the docs. Thank you!

**Comment by smatsson** [comment]
Hi! 

Thanks for the kind words and the great feedback. I can understand that this causes issue with network delays and similar. 

The design behind the different store interfaces is to mimic how the tus protocol is organized. It contains a core protocol (represented by the `ITusStore` interface) and then several extensions (each represented by an interface). A store implements one or more of these interfaces and this is used as feature detection for what the client can use in tusdotnet. The overall architecture of tusdotnet is the split between the protocol handler and the store where the protocol handler takes care of the protocol while the store takes care of everything related to the actual file. In general I think that one would implement several interfaces in the same class, given that the default tus configuration only supports a single store per request, but it's absolutely possible to split it between several classes as well.

To me it sounds like you want some kind of transactional state in your store so that the store knows when to push changes to some external storage. As this is an implementation detail in the store I don't think an event is the correct way of handling this but rather some kind of interface that the store can implement that the protocol handler will call automatically when done or an if an exception occurs. This has been discussed before but nothing amounted from that discussion. Would this solve your use case?

As you probably figured out, the lifetime of a store is undefined by tusdotnet and you could set it to whatever you want and just provide it to the `app.UseTus` method. Adding additional caching inside the store and use a scoped lifetime is possible if you wish to implement it that way. 

You could probably solve this already by abusing the `OnAuthorizeAsync` event to preload data and have an extra middleware just before `app.UseTus` that does the push. Not ideal but absolutely doable:

```csharp

public interface IPreload
{
    Task Preload(string fileId);
}

public interface IPush
{
    Task Push();
}


app.Use(async (context, next) =>
{
    await next();

    if (context.Request.Method is "post" or "patch" or "delete")
    {
        var config = context.RequestServices.GetService<DefaultTusConfiguration>();
        await ((IPush)config.Store).Push();
    }
});

app.UseTus(_ => new()
{
    //Store = new MyNetworkStore(),
    Events = new Events()
    {
        OnAuthorizeAsync = async authorizeContext =>
        {
            // Do normal auth first 

            await ((IPreload)authorizeContext.Store).Preload(authorizeContext.FileId);
        }
    }
});

```

**Comment by mayoatte** [comment]
Thank you, @smatsson . Appreciate all the work and the timely reply.

The choice to mimic the tus protocol makes a lot of sense to me. Also, the suggestion to implement several interfaces in the same class seems natural. Where it breaks down for us is actually the "feature-detection." There are several extensions that we do not want to make available to external / tus clients but want to make use of internally. For example: we do not want to support termination externally, but want to be able to use the termination store internally for management of files (e.g. it's nice to use the termination store for incomplete files because it encapsulates both metadata and the file which are stored in different data stores). Same thing to some extent with expiration. Because these features are auto detected we have no choice to but to implement these stores separately. And we also have to be very careful around which stores we expose on the default configuration. 

I saw some mentions of removing this automatic feature detection in some issues (in favor of explicit configuration), I would like add another voice in support of this. The automatic feature detection (and single store) was very controversial within our team and we spent the most code review time discussing this and making sure to work around it. I think there is some conflation of intent (data stores, extensions supported,  class organization) with the current setup that could be improved by:

- Allowing for configuring the separate stores explicitly (ITusStore, ICreationStore, ITerminationStore etc.)
- Explicit configuration of supported extensions

Using an explicit call to an interface method on the store instead of an event might work. It does depend on which "store" the interface method is on though. In our setup, the creation data, expiration data and metadata are all stored in a single database record (not a blob store). We would want to load that single database record once at the beginning of processing, but that would be difficult because of the class split / feature detection.

Yep, saw how the instances are provided to the middleware and your idea of providing scoped instances makes sense. Same with the additional middleware workaround for pushing state. Will explore those.

Thanks again!

**Comment by smatsson** [comment]
Hi,

Thanks for the detailed explanation. It's so much easier to reason about solutions when one gets all the details :) 

From reading your comment it seems to me like the real problem is the automatic feature detection and I can absolutely see how that would be an issue. It has been proposed before but there hasn't really been enough time for me (or anyone else really, haven't seen a PR ;) ) to implement it. I'm not super enthusiastic about the idea to have multiple store implementations provided to the configuration as I think this will lead to more complex configurations that are hard to follow and frankly I don't see it being used that much. Adding something like "AllowedExtensions" to the configuration object would definitely be possible. Would that solve your issue? 
```csharp

return new DefaultTusConfiguration
            {
                UrlPath = "/files",
                Store = ...
                AllowedExtensions = new[] { 
                    ExtensionConstants.Creation, 
                    ExtensionConstants.CreationWithUpload, 
                    ExtensionConstants.Concatenation 
                },
                Events = ...

```

The store provided would still need to support the features but this would give the developers control of what features are used. It would block both feature detection in the form of the Tus-Extension headers in the OPTIONS request but also prevent tusdotnet from actually performing some operations internally, e.g. setting an expiration date if the expiration extension is not allowed.

**Comment by mayoatte** [comment]
Thanks! Glad the ramblings are helpful :)

The AllowedExtensions idea you have above would alleviate some of the pain and I really appreciate it. At the minimum, it would allow us to:

- Combine some of the store implementations
- Which in turn should unlock being able to store some state in in instances

I like it.

As for PRs, I am interested in contributing here. Don't feel like I know enough yet (our usage of Tus is still in development, so no actual production usage yet) but hope to soon. Specifically, I am interested in some of the discussions around making the library more idiomatic (maybe controllers / minimal APIs, DI etc), would love to add to the effort there once the direction is set.

I appreciate the time. Thank you.

**Comment by smatsson** [comment]
Glad to hear it. I don't have a lot of time right now so hopefully the work around in my first post could get you started. There have been earlier discussion on making some kind of "store proxy" (search for "proxy" in [this issue](https://github.com/tusdotnet/tusdotnet/issues/116)) which would implement all interfaces and use the store provided to it as a way of determining what extensions are available. We could add the list of allowed extensions to this implementation. We probably also want to split the feature detection in two, one for the server (i.e. "what does this store support") and one for the client (i.e. "what extensions are allowed for the client to use").

> As for PRs, I am interested in contributing here. Don't feel like I know enough yet (our usage of Tus is still in development, so no actual production usage yet) but hope to soon.

It was not meant as a jab at you so sorry if it came across that way. It was more of a micro rant on open source in general. :) I absolutely understand the issues with making PRs in repos where one is not familiar with the code and design.

**Comment by mayoatte** [comment]
Yep, the workaround is good enough to get us going, thanks!

I found the mention of the proxy in that issue (fantatisc discussion in there). I am a fan of many of the suggestions in there,  although I am not sure yet about the value of the proxy specifically. In my head:

- The various tus interfaces on the Tus store are sufficient for determining what the server supports internally (backend). The implementation of those interfaces should be able to be injected into other services and used freely on the backend. This seems possible today and we do a bunch of this injection already in our code. 

- The client side feature detection (what does the server expose to the outside world) feels like something that should be configured as part of Tus configuration on startup. The library would never expose the headers for any extensions not explicitly configured here even if the tus store implemented the extensions. There might need to be some validation to make sure that the store implements the configured capabilities.

In this view of the world, the TusStorageService from that threads doesn't add as much value, but i am likely missing something. I am also checking out the three child issues as well, exciting stuff.

`It was not meant as a jab at you so sorry if it came across that way. It was more of a micro rant on open source in general. :)`

I didn't take it as a jab at all, more like I was thinking the same thing about open source and wishing I wasn't doing what most most people do :)

**Comment by smatsson** [comment]
Pardon the late reply, I've been out sick. 

>The various tus interfaces on the Tus store are sufficient for determining what the server supports internally (backend).

In a way I agree with you here. The downside to this design is that you might need to inject multiple interfaces into your service for doing some common tasks. E.g. checking if a file is complete and then reading it would require you to inject both `ITusStore` and `ITusReadableStore`. With the store proxy/adapter/facade/whatever one would only inject one thing. 

@SnGmng has implemented a store adapter in his POC code linked previously. In short: We [grab all supported extensions](https://github.com/SnGmng/tusdotnet/blob/POC/controller-pattern-better-configuration/Source/tusdotnet/ExternalMiddleware/EndpointRouting/TusProtocolHandlerEndpointBased.cs#L100) and then allow [the dev to disable specific extensions](https://github.com/SnGmng/tusdotnet/blob/POC/controller-pattern-better-configuration/Source/tusdotnet/ExternalMiddleware/EndpointRouting/Controllers/TusControllerBase.cs#L151
). Given that we now have the store adapter we can now easily check if we [support something without casting to a specific interface](https://github.com/SnGmng/tusdotnet/blob/POC/controller-pattern-better-configuration/Source/tusdotnet/ExternalMiddleware/EndpointRouting/StorageOperations/WriteOpertionHandler.cs#L99).

**Comment by mayoatte** [comment]
> I've been out sick.

I'm sorry to hear that, hope you're feeling better!

The reasoning above makes sense to me. What's the next step for that POC?

**Comment by smatsson** [comment]
I have started looking at it but haven't really had the time. Feel free to join in the discussions if you like. https://github.com/tusdotnet/tusdotnet/issues/148 :)

**Comment by smatsson** [comment]
Some ongoing work for this issue: https://github.com/tusdotnet/tusdotnet/tree/154-allow-disabling-extensions

Basic idea is that one can enable or disable extensions using the new `AllowedExtension` property of `DefaultTusConfiguration`. The store still needs to support the extension.

Allow all but `termination` and `checksum`:
```
return new DefaultTusConfiguration
    {
        UrlPath = "/files",
        ...
        AllowedExtensions = TusExtensions.All.Except(ExtensionConstants.Termination, ExtensionConstants.Checksum),
```

Allow no extensions at all:
```
return new DefaultTusConfiguration
    {
        UrlPath = "/files",
        ...
        AllowedExtensions = TusExtensions.None,
```

Only allow `creation` (note that this will disable `creation-with-upload`):
```
return new DefaultTusConfiguration
    {
        UrlPath = "/files",
        ...
        AllowedExtensions = new TusExtensions(ExtensionConstants.Creation),
```

**Comment by SebastianStehle** [comment]
@mayoatte How would you implement that with azure blob storage? I got a feature request for my project to implement resumable uploads but none of the file providers that I support, support appending data right now. The only option I see would be to implement something by myself and split up the data into smaller chunks, then download all chucks and concatenate them again before uploading the final file.

I need to support S3, Azure Blob, Google Cloud and GridFS.

**Comment by mayoatte** [comment]
@SebastianStehle 

Do you know what kind of blobs you're dealing with in Azure Blob Storage? If you're dealing with:

**[Append Blobs](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs#about-append-blobs)**: This should integrate well with Tus, especially if you can get the block sizes to line up with Tus. As each Tus call to append happens, you make a similar call to append to the azure blob just forwarding the stream along. Note that you can't modify the blob later though.

**[Block Blobs](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs#about-block-blobs)**: One way to handle these is to treat each Tus append as an uncommited block in the block blob. When the Tus upload is done you can commit the block list (you might have to track this yourself).

In both of these scenarios, it helps to have a database of some kind to help with quick queries (size so far, mapping between the tus upload and the blob etc) which is what led me to posting this issue in the first place.

Hope some of that was helpful. I can try to explain more if you have questions.

**Comment by mayoatte** [comment]
@smatsson 

I just realized i missed your last post about ongoing work on disabling extensions. Apologies, I was on vacation then and missed the notification.

Thanks a ton for working on this. We will definitely try to use it and i will report back on how it goes. How is everything else going, got out of touch on the next version progress in the last month.

**Comment by SebastianStehle** [comment]
Thanks @mayoatte 

Because I need to support so many storages my idea is the following:

1. Each append is its own block blob or S3 object and so on.
2. All metadata is stored in a database. Basically I just need a key-value-table for that, because all queries are by fileId.
3. When all parts are uploaded, I create a local temporary file and download the blobs one by one to this temporary file. I need this file anyway, because the file is needed for metadata detection (file type and so on).

With this approach my storage interface only needs two methods:

```
IStorage {
  Task<long> UploadAsync(string id, Stream stream);
  Task DownloadAsync(string id, Stream stream);
  Task DeleteAsync(string id);
}
```

**Comment by mayoatte** [comment]
Anytime, @SebastianStehle 

You've got the general idea. I do think it might be more complicated than the interface above might suggest. There will likely be nuances in the different stores that will muddy up the interface. Another option might be to have an S3TusStore, AzureBlobTusStore etc. More work for sure, but might be able to handle those nuances.

**Comment by SebastianStehle** [comment]
I tried this approach: https://github.com/Squidex/assets/blob/tus-adapter/Squidex.Assets.TusAdapter/AssetTusStore.cs

It is not tested, I just wanted to figure out how easy it would be to implement this. There are 2 major things left to do:

1. Change my IAssetStore interface to return the number of written bytes.
2. Test how each adapter behaves when the request is cancelled.

And of course testing, testing, testing.

**Comment by SebastianStehle** [comment]
So far it works great. I have implementation for 

Google Cloud, Azure S3, Amazon Blob Storage, MongoDB GridFS, Folders and FTP now :)

**Comment by smatsson** [comment]
@SebastianStehle That is super cool. Thanks for sharing! If you decide to distribute the store implementations by themselves please let me know and I'll add a link from the readme.

@mayoatte No problem at all! I have been out sick so I haven't gotten around to releasing a new version anyway :)

**Comment by SebastianStehle** [comment]
I have built them for Squidex (my headless CMS), but I also use them in a separate project. Therefore they are on nuget. But the implementation is not perfect yet. I need more testing.

**Comment by smatsson** [comment]
Just le me know when/if you want it added. No hurries :)

**Comment by SebastianStehle** [comment]
I have tested it a lot recently and I would say that the store works good enough now.

**Comment by mayoatte** [comment]
@SebastianStehle 

Glad it's working! Also that's incredible productivity to get so many implementations done so fast. 

Btw: I tried to follow the GCS flow a bit but couldn't understand where the parts of the file were combined.

**Comment by smatsson** [comment]
@SebastianStehle Very nice indeed! Can the stores be used outside your CMS and if so, would you like me to add a link to the tusdotnet readme? I can see it being mutually beneficial to both our projects

@mayoatte Sorry for the long delay but I have finally published tusdotnet 2.6 which has support for disabling/enabling extensions manually. Release notes: https://github.com/tusdotnet/tusdotnet/releases/tag/2.6

**Comment by SebastianStehle** [comment]
Yes, 
there is no dependency and I also use it for other projects.

But you have to write a custom Metadata store. There is only a MongoDB implementation right now.

**Comment by mayoatte** [comment]
@smatsson Thanks a ton for the release! Will upgrade and report back how it goes.
